src/Compilers/CSharp/Test/Emit3/Attributes/AttributeTests_Embedded.cs

```diff
@@ -193,61 +195,400 @@ public static void Main()
+                        public EmbeddedAttribute() { }
+                    }
+                }
+                class Test
+                {
+                    public void M(in int p)
+                    {
+                        // This would trigger generating another EmbeddedAttribute, if we hadn't already defined one
+                    }
+                }
+                """;
+
+            CreateCompilation(code, assemblyName: "testModule", targetFramework: TargetFramework.NetStandard20).VerifyEmitDiagnostics(
+                // (4,20): error CS9271: The type 'Microsoft.CodeAnalysis.EmbeddedAttribute' must be non-generic, internal, sealed, non-static, have a parameterless constructor, inherit from System.Attribute, and be able to be applied to any type.
+                //     internal class EmbeddedAttribute : System.Attribute
+                Diagnostic(ErrorCode.ERR_ReservedTypeMustFollowPattern, "EmbeddedAttribute").WithArguments("Microsoft.CodeAnalysis.EmbeddedAttribute").WithLocation(4, 20));
         }
 
         [Fact]
-        public void EmbeddedAttributeInReferencedModuleShouldTriggerAnErrorIfCompilerNeedsToGenerateOne()
+        public void EmbeddedAttributeFromSourceValidation_File()
         {
-            var module = CreateCompilation(options: TestOptions.ReleaseModule, assemblyName: "testModule", source: @"
-namespace Microsoft.CodeAnalysis
-{
-    public class EmbeddedAttribute : System.Attribute { }
-}");
+            var code = """
+                namespace Microsoft.CodeAnalysis
+                {
+                    [Embedded]
+                    file sealed class EmbeddedAttribute : System.Attribute
+                    {
+                        public EmbeddedAttribute() { }
+                    }
+                }
+                class Test
+                {
+                    public void M(in int p)
+                    {
+                        // This would trigger generating another EmbeddedAttribute, if we hadn't already defined one
+                    }
+                }
+                """;
+
+            var verifier = CompileAndVerify(code, targetFramework: TargetFramework.NetStandard20, symbolValidator: module =>
+            {
+                var embeddedAttribute = module.ContainingAssembly.GetTypeByMetadataName(AttributeDescription.CodeAnalysisEmbeddedAttribute.FullName);
+                Assert.NotNull(embeddedAttribute);
+                Assert.Equal("Microsoft.CodeAnalysis.EmbeddedAttribute", embeddedAttribute.ToTestDisplayString());
+
+                var embeddedAttributeFile = module.ContainingAssembly.GetTypeByMetadataName(AttributeDescription.CodeAnalysisEmbeddedAttribute.FullName + "@<unknown>");
+                Assert.NotNull(embeddedAttribute);
+                Assert.Equal("Microsoft.CodeAnalysis.EmbeddedAttribute", embeddedAttribute.ToTestDisplayString());
+                Assert.NotSame(embeddedAttribute, embeddedAttributeFile);
+            });
```

Comment:
>} [](http://example.com/codeflow?start=12&length=1)

Consider verifying the `[Embedded]` attribute applied to the `file` type is `<unknown>_EmbeddedAttribute`.